login.go

```diff
@@ -0,0 +1,202 @@
+	}, nil
+}
+
+func getDeviceCode(ctx context.Context, client *ConductoroneAPI, clientID string) (*DeviceCodeResponse, error) {
+	httpClient := client.sdkConfiguration.DefaultClient
+
+	deviceCodeURL := "https://" + client.sdkConfiguration.ServerURL + "/auth/v1/device_authorization"
+	vals := url.Values{}
+	vals.Add("client_id", clientID)
+
+	req, err := http.NewRequestWithContext(ctx, "POST", deviceCodeURL, strings.NewReader(vals.Encode()))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	data, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	codeResp := &DeviceCodeResponse{}
+	err = json.Unmarshal(data, codeResp)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Printf("Open: %s\n", codeResp.VerificationURI)
+	return codeResp, nil
+}
+
+func doTokenRequest(ctx context.Context, client *ConductoroneAPI, clientID string, deviceCodeResp *DeviceCodeResponse) (*tokenResponse, error) {
+	httpClient := client.sdkConfiguration.DefaultClient
+
+	tokenURL := "https://" + client.sdkConfiguration.ServerURL + "/auth/v1/token"
+	vals := url.Values{}
+	vals.Add("client_id", clientID)
+	vals.Add("device_code", deviceCodeResp.DeviceCode)
+	vals.Add("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
+
+	startLoop := time.Now()
+	for {
+		select {
+		case <-time.After(time.Duration(deviceCodeResp.Interval) * time.Second):
+			if time.Now().After(startLoop.Add(time.Duration(deviceCodeResp.ExpiresIn) * time.Second)) {
+				return nil, errors.New("timeout")
+			}
+		case <-ctx.Done():
+			return nil, nil
+		}
+
+		req, err := http.NewRequestWithContext(ctx, "POST", tokenURL, strings.NewReader(vals.Encode()))
+		if err != nil {
+			return nil, err
+		}
+		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+		resp, err := httpClient.Do(req)
+		if err != nil {
+			return nil, err
+		}
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		if resp.StatusCode >= 300 {
+			errResp := &oauth2Error{}
+			err = json.Unmarshal(body, errResp)
+			if err != nil {
+				return nil, err
+			}
+
+			if errResp.ErrorType == "authorization_pending" {
+				continue
+			}
+
+			return nil, errors.New(errResp.ErrorDescription)
+		}
+
+		tokenResp := &tokenResponse{}
+		err = json.Unmarshal(body, tokenResp)
+		if err != nil {
+			return nil, err
+		}
+
+		return tokenResp, nil
+	}
+}
+
+func doClientCredentialRequest(ctx context.Context, client *ConductoroneAPI, tokenResp *tokenResponse) (*clientResp, error) {
+	httpClient := client.sdkConfiguration.DefaultClient
+	pccURL := "https://" + client.sdkConfiguration.ServerURL + "/api/v1/iam/personal_clients"
+
+	personalClientReq, err := http.NewRequestWithContext(ctx, "POST", pccURL, bytes.NewReader([]byte("{\"display_name\": \"Created From Cone\"}")))
```

Comment:
this string should probably be a parameter to the Login func?

Reply:
Yeah good call